src/autotype/PickcharsDialog.cpp

```diff
@@ -0,0 +1,178 @@
+    }
+
+    // Places the window on the active (virtual) desktop instead of where the main window is.
+    setAttribute(Qt::WA_X11BypassTransientForHint);
+    setWindowFlags((windowFlags() | Qt::WindowStaysOnTopHint | Qt::MSWindowsFixedSizeDialogHint)
+                   & ~Qt::WindowContextHelpButtonHint);
+    setWindowIcon(icons()->applicationIcon());
+
+    m_ui->setupUi(this);
+
+    int row = 0, col = 0;
+    int width = string.length() >= 100 ? 20 : string.length() >= 60 ? 15 : 10;
+    for (const auto& ch : string) {
+        auto btn = new QPushButton(QString::number(row * width + col + 1));
+        btn->setProperty("char", ch);
+        btn->setProperty("gridpos", QPoint(row, col));
+        connect(btn, &QPushButton::clicked, this, &PickcharsDialog::charSelected);
+        m_ui->charsGrid->addWidget(btn, row, col);
+        if (++col >= width) {
+            col = 0;
+            ++row;
+        }
+    }
+    // Prevent stretched buttons
+    if (row == 0 && col <= 5) {
+        m_ui->charsGrid->addItem(new QSpacerItem(5, 5, QSizePolicy::MinimumExpanding), row, col);
+    }
+    m_ui->charsGrid->itemAtPosition(0, 0)->widget()->setFocus();
+
+    connect(m_ui->buttonBox, &QDialogButtonBox::accepted, this, &QDialog::accept);
+    connect(m_ui->buttonBox, &QDialogButtonBox::rejected, this, &QDialog::reject);
+
+    // Navigate grid layout using up/down/left/right motion
+    new QShortcut(Qt::Key_Up, this, SLOT(upPressed()));
+    new QShortcut(Qt::Key_Down, this, SLOT(downPressed()));
+    // Remove last selected character
+    auto shortcut = new QShortcut(Qt::Key_Backspace, this);
+    connect(shortcut, &QShortcut::activated, this, [this] {
+        auto text = m_ui->selectedChars->text();
+        m_ui->selectedChars->setText(text.left(text.size() - 1));
+    });
+    // Submit the form
+    shortcut = new QShortcut(Qt::CTRL + Qt::Key_S, this);
+    connect(shortcut, &QShortcut::activated, this, [this] { accept(); });
+}
+
+void PickcharsDialog::upPressed()
+{
+    auto focus = focusWidget();
+    if (!focus) {
+        return;
+    }
+
+    auto gridpos = focus->property("gridpos");
+    if (gridpos.isValid()) {
+        auto item = m_ui->charsGrid->itemAtPosition(gridpos.toPoint().x() - 1, gridpos.toPoint().y());
+        if (item) {
+            item->widget()->setFocus();
+        } else {
+            // Cannot go up, try item to left
+            item = m_ui->charsGrid->itemAtPosition(gridpos.toPoint().x(), gridpos.toPoint().y() - 1);
+            if (item) {
+                item->widget()->setFocus();
+            } else {
+                m_ui->pressTab->setFocus();
+            }
+        }
+    } else if (focus == m_ui->selectedChars) {
+        // Move back to the last button
+        auto item = m_ui->charsGrid->itemAt(m_ui->charsGrid->count() - 1);
+        if (item) {
+            item->widget()->setFocus();
+        }
+    } else if (focus == m_ui->pressTab) {
+        m_ui->selectedChars->setFocus();
+    }

```

Comment:
No it wouldn't because I need to probe to see if we exceeded the width anyway which is what the first item does.

Reply:
With a modulo operation, you would simply wrap around, no need to check anything. That's IMHO less confusing than switching to left/right movement.

Reply:
OK I implemented your recommendation and removed the alt-logic of going left/right even though you press Up/Down. Now it will stop at the top and move straight to the line edit at the bottom of the grid.